### ISSUE TITLE
[REVIEW]: Precise calculations on hypercomplex numbers in C++

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#3096]
**Submitting author:** @AngryMaciek (<a href="http://orcid.org/0000-0003-1361-7301">Maciej Bak</a>)
**Repository:** <a href="https://github.com/AngryMaciek/hypercomplex" target ="_blank">https://github.com/AngryMaciek/hypercomplex</a>
**Version:** v1.0.0
**Editor:** @melissawm
**Reviewers:** @dvberkel, @nnadeau
**Archive:** Pending

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/c5644100d52193f0379609d553ece4a8/status.svg)](https://joss.theoj.org/papers/c5644100d52193f0379609d553ece4a8)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/c5644100d52193f0379609d553ece4a8"><img src="https://joss.theoj.org/papers/c5644100d52193f0379609d553ece4a8/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/c5644100d52193f0379609d553ece4a8/status.svg)](https://joss.theoj.org/papers/c5644100d52193f0379609d553ece4a8)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/AngryMaciek/hypercomplex" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@dvberkel & @bonh, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @melissawm know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/AngryMaciek/hypercomplex">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@AngryMaciek) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/AngryMaciek/hypercomplex">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@AngryMaciek) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [ ] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [ ] **Installation:** Does installation proceed as outlined in the documentation?
- [ ] **Functionality:** Have the functional claims of the software been confirmed?
- [ ] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [ ] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [ ] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [ ] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [ ] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [ ] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [ ] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [ ] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [ ] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [ ] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [ ] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [ ] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/AngryMaciek/hypercomplex">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@AngryMaciek) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [ ] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [ ] **Installation:** Does installation proceed as outlined in the documentation?
- [ ] **Functionality:** Have the functional claims of the software been confirmed?
- [ ] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [ ] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [ ] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [ ] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [ ] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [ ] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [ ] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [ ] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [ ] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [ ] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [ ] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [ ] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.88  T=0.34 s (71.4 files/s, 63096.0 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
C/C++ Header                     2           3233           1408          14285
C++                              4             89             66           1249
YAML                             6             66              0            243
Markdown                         8             71              0            233
XML                              1              2              4             98
TeX                              1              7              0             75
make                             1             17             31             35
HTML                             1              0              0              6
-------------------------------------------------------------------------------
SUM:                            24           3485           1509          16224
-------------------------------------------------------------------------------


Statistical information for the repository '7e57fef42a3fd14af9551445' was
gathered on 2021/03/10.
The following historical commit information, by author, was found:

Author                     Commits    Insertions      Deletions    % of changes
Maciej Bak                     205         40007          21364           94.88
Maciek Bąk                       6          2501            814            5.12

Below are the number of rows from each author that have survived and are still
intact in the current revision:

Author                     Rows      Stability          Age       % in comments
Maciej Bak                20330           50.8          3.2                7.27
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- None

MISSING DOIs

- 10.1007/978-3-7643-7791-5_2 may be a valid DOI for title: Quaternions
- 10.1007/978-1-4419-7719-9_6 may be a valid DOI for title: Boost C++ libraries

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03096/joss.03096/10.21105.joss.03096.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03096/joss.03096/10.21105.joss.03096.pdf) :page_facing_up: :point_left:

**Comment by dvberkel** [comment]
@melissawm I finished my review of the paper. In my opinion the software presented is of excellent quality and in my book ticks all the required marks.
Furthermore, the software goes above and beyond to provide a good library.

**Comment by melissawm** [comment]
Thank you, @dvberkel !

**Comment by bonh** [comment]
@melissawm I am somehow unable to tick the boxes in my list above. Did I forget to accept some invitation (stupid me...)?

**Comment by bonh** [comment]
Hi @AngryMaciek ! 

Great software and super clean engineering and project management! I have a few comments, suggestions and so on. I will start to open issues in your repo so that we can discuss them.

**Comment by bonh** [comment]
@melissawm I suggest that you get into that one #AngryMaciek/hypercomplex#26. Perhaps I am misreading the guidelines of JOSS.

**Comment by whedon** [comment]
:wave: @bonh, please update us on how your review is going (this is an automated reminder).

**Comment by melissawm** [comment]
@bonh don't worry about whedon, it's an automatic setting :)

**Comment by bonh** [comment]
@melissawm still cannot tick the boxes. Am i doing something wrong?

**Comment by melissawm** [comment]
Sorry - let me re-invite you

**Comment by melissawm** [comment]
@whedon re-invite @bonh  as reviewer

**Comment by whedon** [comment]
OK, the reviewer has been re-invited.

@bonh  please accept the invite by clicking this link: https://github.com/openjournals/joss-reviews/invitations

**Comment by bonh** [comment]
@melissawm based on #AngryMaciek/hypercomplex#26 I flag this submission for scope check.

**Comment by bonh** [comment]
@melissawm I will continue my review on your sign. However, If you find me unsuitable as a reviewer (based on missing experience and knowledge in the specific field, see #AngryMaciek/hypercomplex#26), I will gladly withdraw from the process.

**Comment by melissawm** [comment]
@bonh @AngryMaciek based on discussions with the JOSS editors, since one reviewer has expressed approval of the submission, we feel the best course of action would be to invite a third reviewer with more familiarity with this field in order to complete the review. In that case, @bonh would be removed as a reviewer. Let me know if you agree.

**Comment by AngryMaciek** [comment]
I am perfectly fine with such resolution.

**Comment by bonh** [comment]
@melissawm I hope that you are not waiting for my reply - I reacted with a thumps up right after your message (and proposed to withdraw in my message above yours). Just asking because this review thread lingers around in my todo list without anything happening.

**Comment by melissawm** [comment]
Ah - this has gotten me a few times, indeed I do not get a notification when that happens. Thanks for the heads up, @bonh !

**Comment by bonh** [comment]
Ohhh no, I wasn’t aware of that!

**Comment by melissawm** [comment]
Hello, @iljah! Would you be willing to review this submission for JOSS? We carry out our checklist-driven reviews here in GitHub issues and follow these guidelines: https://joss.readthedocs.io/en/latest/review_criteria.html

You can check a bit of the history of this review here in the thread, let me know if you have any questions or need more information.

**Comment by iljah** [comment]
@melissawm thanks for the opportunity but I have very little experience even with regular complex numbers and no experience with more complex numbers. @AngryMaciek what did stand out from documentation (https://angrymaciek.github.io/hypercomplex/html/index.html) was need for manual memory management when using MPFR types. I've almost not used MPFR but at least it should be possible to automate much of what is done manually at end of above documentation. Following notice is also unusual for header-only c++ library: "Please do not mix this library with another MPFR-dependant code in the same translation unit.".

**Comment by AngryMaciek** [comment]
Thanks for the comment @iljah!
1.  Each _MPFR_  variable has to be manually freed from the memory, that's how the library is designed. In the specialisation for `Hypercomplex` which is based on these types I do free the internal variables in the [class destructror](https://github.com/AngryMaciek/hypercomplex/blob/2f7db6fdc4dbeff9b5115a8244f4d7c8819b162f/hypercomplex/Hypercomplex.hpp#L652). These commands you probably refer to, at the end of the documentation, are related to freeing memory from an (mock) array which user could create in their main program.
2. 'Non-mixing' information is there because of how _MPFR_ stores precision related settings in global variables, but maybe the notice is obsolete now. I will have to check this once again in the documentation...

**Comment by iljah** [comment]
I mean that at least code shown in documentation could be moved to hypercomplex constructor I think. Hypercomplex<mpfr_t, 8> already knows that there will be 8 mpfr types so it could call mpfr_init2 in constructor. Constructor could also call mpfr_set_d which would allow using H{1,2,3,...} directly. Although now I notice that same thing applies to doubles so hypercomplex never owns its storage? That's surprising since all sizes are known at compile time.

**Comment by AngryMaciek** [comment]
@iljah OK thanks for the input! I remember that there were some constraints on how the constructor would work due to the recursive implementation of the multiplication operator. I will for sure look into this deeper once we move on with the review.

**Comment by melissawm** [comment]
Hi, @iljah. I understand you're not available to review, is that correct?

**Comment by melissawm** [comment]
Hello, @nnadeau - would you be willing to review this submission for JOSS? We carry out our checklist-driven reviews here in GitHub issues and follow these guidelines: https://joss.readthedocs.io/en/latest/review_criteria.html

You can check a bit of the history of this review here in the thread. Let me know if you have any questions or need more information.

**Comment by iljah** [comment]
@melissawm correct, I probably couldn't contribute any more than I already have.

**Comment by nnadeau** [comment]
@melissawm sure, i can be available. "hypercomplex numbers" are outside of my direct specialty, but I can definitely review the C++ and the overall scholarly effort.

**Comment by melissawm** [comment]
Sounds great! I´ll add you as a reviewer and will edit the checklist above to include your review. Thank you!

**Comment by melissawm** [comment]
@whedon add @nnadeau as reviewer

**Comment by whedon** [comment]
OK, @nnadeau is now a reviewer

**Comment by bonh** [comment]
@whedon remove @bonh as reviewer

**Comment by whedon** [comment]
I'm sorry @bonh, I'm afraid I can't do that. That's something only editors are allowed to do.

**Comment by bonh** [comment]
> I'm sorry @bonh, I'm afraid I can't do that. That's something only editors are allowed to do.

Hahaha, I though I was acting clever here... :+1: 
@melissawm Please be so kind!

**Comment by melissawm** [comment]
@whedon remove @bohn as reviewer

**Comment by whedon** [comment]
OK, @bohn is no longer a reviewer

**Comment by melissawm** [comment]
Thank you for all your inputs and thoughtful considerations, @bohn!

**Comment by melissawm** [comment]
@whedon remove @bonh as reviewer

**Comment by whedon** [comment]
OK, @bonh is no longer a reviewer

**Comment by nnadeau** [comment]
@melissawm @AngryMaciek , this is a great tool, but I cannot recommend publication based on the grounds of [`Substantial scholarly effort`](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

- Only represents about 2 months of work ([judging by commit history](https://github.com/AngryMaciek/hypercomplex/graphs/contributors?from=2020-10-18&to=2020-12-31&type=c))
- Only <1000 lines of code, much of which is just C++ templating

```
~/git/hypercomplex ‹master› » cloc hypercomplex                                       2 ↵
       2 text files.
       2 unique files.
       0 files ignored.

github.com/AlDanial/cloc v 1.82  T=0.01 s (211.4 files/s, 125121.1 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
C/C++ Header                     1             95            338            668
make                             1             17             31             35
-------------------------------------------------------------------------------
SUM:                             2            112            369            703
-------------------------------------------------------------------------------
```

- The `Statement of need` does not convince me **why** I should choose to use this library vs simply writing my own small utility functions to handle this use case. Including 3rd party packages is a risk and there has to be a significant value proposition to justify the risk.
- Yes, Boost is bloated and doesn't include this functionality, but this tool would make the most positive impact on the world if it was contributed to the Boost project

**Comment by melissawm** [comment]
Hello all, based on the discussion I'm flagging this for the EiC on rotation to look at. @AngryMaciek if you feel like you have anything to add, please do. Let me know if I can answer any further questions. Thanks!

**Comment by melissawm** [comment]
@whedon query scope

**Comment by AngryMaciek** [comment]
Dear @melissawm,

Yes, of course! I would very much like to respond to the points raised by @nnadeau:

[1] I am not sure why, but that contribution graphs do not reflect the commit history accurately. Please take a look at the [commit history](https://github.com/AngryMaciek/hypercomplex/commits/master) of `master` directly. Initial commit for this repo dates back to 22 October, the last commit is from 16 February. That would result in not 2 months but official 3 months and 3 weeks of work (so, almost 2x increase). Of course it's not only the number of days that matters, but I am just replying to the reviewers comment. According to the explicit specifications in the [JOSS Submission guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort), this should be fine:
> As a rule of thumb, JOSS’ minimum allowable contribution should represent not less than three months of work for an individual.

[2] The statistics which the reviewer posted above are correct, the core template file is not big. However, I believe that we should stick to counting the lines of code just as _JOSS_ does, i.e. please look at [this post](https://github.com/openjournals/joss-reviews/issues/3096#issuecomment-795850551) by @whedon. `cloc` is configured here to explicitly count formats such as XML, YAML, Markdown, HTML etc. These are not pure code but their reflect my work and contribute to the overall high state of the library from the software engineering perspective (CI, documentation, READMEs, tests). Personally, it is not a surprise for me that @whedon is configured here in such manner - after all this journal emphasizes the high-quality software engineering therefore it seems reasonable that it counts files other then the core code.
Please note that inside this repository I also include [Catch2](https://github.com/catchorg/Catch2) testing library, so we need to deduct it from the `cloc` output:
```
$ cloc hypercomplex/.test/unit/catch.hpp
       1 text file.
       1 unique file.
       0 files ignored.

github.com/AlDanial/cloc v 1.88  T=0.07 s (15.3 files/s, 272029.4 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
C/C++ Header                     1           3138           1070          13617
-------------------------------------------------------------------------------
```
Therefore, in the end we are left with `16224 - 13617 = 2607` lines of code for the whole repository, as counting by the _JOSS_ standards. Again, quoting the submission instructions: such number should not raise concerns regarding the repository size:
> Total lines of code (LOC). Submissions under 1000 LOC will usually be flagged, those under 300 LOC will be desk rejected.

[3] I cannot really say I agree with this mindset. We are having here a _Journal of Open Source Software_ which is - as it states - mostly focused on the quality of the software, CI/CD testing, documentation etc. This is exactly so that not everyone would need to write custom functions (and their unit tests!) from scratch. This is one of the main ideas behind open source software - so that a large group of people share small building blocks. In that way we can collectively build bigger and greater things. Modularisation is a key feature in the process.
Additionally, about the risk - I claim that it is quite the opposite. Including a high-quality library with its own testing framework might be less of a risk than relying on a custom piece of code which was written on-the-fly.
I cannot speak for others than myself but generally when I see a solid library out there I am happy to use it instead of reinventing the wheel myself. And the submission guidelines for _JOSS_ enforce that the libraries published here *are* solid (so that other people can indeed rely on my work).

[4] OK, I agree here. If you believe it makes sense I will reach out to _Boost_ developers and ask if they would be interested in including _Hypercomplex_(?)

**Comment by bonh** [comment]
Hi @melissawm, 
the submission is still flagged for editorial review - so it is not done yet and I feel uncomfortable to start another review while that one still lingers around (even if I'm not a part of it anymore). My point is, that I might not have the heart to ask for another scope check (or some other editorial stuff) during the review of another submission even if I feel it might be needed - after all we all do this on the side and I greatly appreciate the work of the editors! I hope you get what I mean and please let me know if I'm stepping on someone's toes here (if so I am sorry!).

**Comment by danielskatz** [comment]
@melissawm - to flag the AiEC, please mention @openjournals/joss-eics Doing a query scope command is generally intended for pre-review and doesn't get tracked well when done by an editor during a review.

**Comment by AngryMaciek** [comment]
Dear @melissawm, @danielskatz and @bonh,

I kindly ask to put a hold on any actions regarding this review. I am in the middle of expanding the repository a little in order to provide a concrete real-world application in the field of lattice-based cryptography. This would significantly elevate the quality and usability of my work. As I am rather busy lately it should take me approx. 1 month to push an updated version to GitHub. Could we please add a `Paused` label to this thread? I will let you know once I'm done and you may proceed then.

Kind Regards,

**Comment by kyleniemeyer** [comment]
@AngryMaciek sure, we can certainly pause your submission for now.

**Comment by melissawm** [comment]
Noted @danielskatz , sorry about that.

**Comment by danielskatz** [comment]
@AngryMaciek - Do you have an update on this?

**Comment by AngryMaciek** [comment]
@danielskatz 
Yes, sorry for the delay. It took me much longer then expected since I needed to learn more theory about advanced cryptography, before even having started to code it. I am essentially adding here a class specialisation to include the functionality described [in this paper](https://eprint.iacr.org/2009/386.pdf) I think it fits perfectly as a concrete application of a general algebraic framework presented in this repo.
I have a strong deadline at work: until the end of this month I am super busy with everyday tasks. I will be able to come back here and finish this off in November.

**Comment by AngryMaciek** [comment]
Dear @danielskatz and @melissawm,
I still need more time to implement the changes I envison for this library.
Maybe it would be better then if I resubmit to _JOSS_ once everything is finished?

**Comment by danielskatz** [comment]
That seems reasonable to me.  @melissawm ?

**Comment by melissawm** [comment]
That sounds good to me, thanks for letting us know @AngryMaciek !

**Comment by kthyng** [comment]
Ok @AngryMaciek I will withdraw your submission. When you are able to resubmit, please link to this partial review to provide context. Thank you to reviewers @dvberkel, @nnadeau for your work on this submission.

**Comment by kthyng** [comment]
@whendon withdraw

**Comment by kthyng** [comment]
@whedon withdraw

**Comment by whedon** [comment]
Paper withdrawn.
